Give CLI args higher priority over environment.

pipejakob opened the pull request:
As written, inherited environment variables like `E2E_UP` were overriding their CLI counterparts like `--up` in the `kubernetes_e2e` scenario. This change reorders the execution so that these command-line flags override environment variables, which seems more correct.

This explains why the kubeadm e2e job is [still running tests](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-gce/587), even though `--test=false` is [given on the command-line](https://github.com/kubernetes/test-infra/blob/535e32a840ee72bde11a8bd43a25728feb71291c/jobs/config.json#L517).

krzyzacy commented:
why `E2E_UP` is in the local env? because of https://github.com/kubernetes/test-infra/blob/master/jenkins/e2e-image/Dockerfile#L22-L24?

This is the behavior we want - user can have local env overwrites scenario envs.

krzyzacy commented:
Maybe it's better to get rid of those from the Dockerfile and make a new image.

pipejakob commented:
Yes, the default values are coming in from the Docker image. That means that if I run a job from within the `kubekins-e2e` image, then arguments like `--up=false` or `--down=false` will always be ignored, because `E2E_UP` and `E2E_DOWN` are already set to true in the environment.

Isn't it a standard pattern to have CLI arguments override their environment counterparts? I don't think I've ever seen another tool where setting an environment variable overrides the argument you supply on the command line.

pipejakob commented:
I should say that I assumed I was fixing an unknown bug. If this is intentionally the desired behavior, then that confuses me, but I can start with a proposal to change the expected behavior _before_ implementing it. :-)

krzyzacy commented:
Hummmm it works in docker mode, I guess `docker run -e` will overwrite the env from the image. However local mode runs from the image already, that's the difference.

Can you also move rest of that block down as well?

pipejakob commented:
Ah, I didn't notice that `cluster` was also coming from CLI args, so you're right that it should probably move down. It looks like a few of these provide defaults that are meant to be overridden by the environment, so I think they'll need to stay. I'll also loop in Erick (who has a TODO right here) to make sure I'm not breaking any assumptions/requirements of these vars, since it looks like inheriting from the environment might be going away anyway.

pipejakob commented:
Okay, I've moved the rest of the block down (except for the default values) and added comments to make the intent of each block clear.

@fejta What do you think about this? You have a TODO on line 359 for removing the environment inheritance altogether, so I assume you have the most context about the needs here. Is this change in the right direction, or will it just break desired behavior?

pipejakob commented:
Ah, looks like I tripped a pylint check (main has gotten too big, and probably needs to be refactored). I want to make sure this is actually headed in the right direction first, then I'll work on appeasing the pylint checks.

krzyzacy commented:
I think this is fine. Feel free to move the entire block down so you won't get the pylint error.

pipejakob commented:
Done.

krzyzacy commented:
looks like now there's a big `Update branch` button you'll need to hit

pipejakob commented:
Rebased!

krzyzacy commented:
/lgtm
